x-pack/plugin/enrich/src/main/java/org/elasticsearch/xpack/enrich/action/InternalExecutePolicyAction.java

```diff
@@ -0,0 +1,170 @@
+import java.util.Set;
+import java.util.concurrent.atomic.AtomicInteger;
+
+import static org.elasticsearch.xpack.core.enrich.action.ExecuteEnrichPolicyAction.Request;
+import static org.elasticsearch.xpack.core.enrich.action.ExecuteEnrichPolicyAction.Response;
+import static org.elasticsearch.xpack.enrich.EnrichPolicyExecutor.TASK_ACTION;
+
+/**
+ * An internal action runs {@link org.elasticsearch.xpack.enrich.EnrichPolicyRunner} and ensures that:
+ * <ul>
+ *     <li>In case the cluster has more than one node, the policy runner isn't executed on the elected master
+ *     <li>Additionally, if the cluster has master only nodes then the policy runner isn't executed on these nodes.
+ * </ul>
+ *
+ * The {@link TransportExecuteEnrichPolicyAction} is a transport action that runs on the elected master node and
+ * the actual policy execution may be heavy for the elected master node.
+ * Although {@link org.elasticsearch.xpack.enrich.EnrichPolicyRunner} doesn't do heavy operations, the coordination
+ * of certain operations may have a non-negligible overhead (for example the coordination of the reindex step).
+ */
+public class InternalExecutePolicyAction extends ActionType<Response> {
+
+    private static final Logger LOGGER = LogManager.getLogger(InternalExecutePolicyAction.class);
+    public static final InternalExecutePolicyAction INSTANCE = new InternalExecutePolicyAction();
+    public static final String NAME = "cluster:admin/xpack/enrich/internal_execute";
+
+    private InternalExecutePolicyAction() {
+        super(NAME, Response::new);
+    }
+
+    public static class Transport extends HandledTransportAction<Request, Response> {
+
+        private final ClusterService clusterService;
+        private final TransportService transportService;
+        private final EnrichPolicyExecutor policyExecutor;
+        private final AtomicInteger nodeGenerator = new AtomicInteger(Randomness.get().nextInt());
+
+        @Inject
+        public Transport(
+            TransportService transportService,
+            ActionFilters actionFilters,
+            ClusterService clusterService,
+            EnrichPolicyExecutor policyExecutor
+        ) {
+            super(NAME, transportService, actionFilters, Request::new);
+            this.clusterService = clusterService;
+            this.transportService = transportService;
+            this.policyExecutor = policyExecutor;
+        }
+
+        @Override
+        protected void doExecute(Task transportTask, Request request, ActionListener<Response> actionListener) {
+            var clusterState = clusterService.state();
+            var node = selectNodeForPolicyExecution(clusterState.nodes());
+            if (clusterState.nodes().getLocalNode().equals(node) == false) {
+                var handler = new ActionListenerResponseHandler<>(actionListener, Response::new);
+                transportService.sendRequest(node, NAME, request, handler);
+                return;
+            }
```

Comment:
Yes, your analysis is right here. It can do more hops than it should do.  I think this can be addressed by changing `selectNodeForPolicyExecution(...)` method the return local node if current node is not elected master node. This should work, since the first time this internal action is invoked is always on the elected master node. The second time this logic is executed on another node then checks checks prevents this logic from selecting yet another node.

Or we can make it more explicit by adding a node id field (of type string) to the request class of this internal action. This would be set when this internal action is executed on the elected master node then on the target node it will always execute (having a check something like this: `if (clusterState.nodes().getLocalNode().id().equals(request.targetNodeId()) {`.

Reply:
I think the core of the issue here (We don't want this action running on master) is solved simplest by that first suggestion. As long as the preconditions for which node it is on are met then it can just execute as normal.

Reply:
Adjusted the logic: f25f1e07d8305b246bc1f55a930e30aff29c09d2